java/fury-core/src/main/java/org/apache/fury/Fury.java

```diff
@@ -310,12 +310,13 @@ private void write(MemoryBuffer buffer, Object obj) {
       buffer.writeInt32(-1); // preserve 4-byte for nativeObjects start offsets.
     }
     // reduce caller stack
-    if (!refResolver.writeRefOrNull(buffer, obj)) {
+    boolean isNull = refResolver.writeRefOrNull(buffer, obj);
+    if (!isNull) {
       ClassInfo classInfo = classResolver.getOrUpdateClassInfo(obj.getClass());
       classResolver.writeClass(buffer, classInfo);
       writeData(buffer, classInfo, obj);
     }
-    if (shareMeta) {
+    if (shareMeta && !isNull) {
```

Comment:
The deserialization will still read class meta, but the offset will be -1, then the read will fail. Could you add an UT for such case?

Reply:
Done.

CI failure is not related to the current changes, fix it in https://github.com/apache/fury/pull/1691.